### PR TITLE
Improve sharing of trees when changing project configuration

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs [hunk 1]
```diff
@@ -862,8 +862,10 @@ public async Task ChangingLanguageVersionReparses()
 
         [Theory]
         [InlineData("#if DEBUG", false, LanguageNames.CSharp)]
+        [InlineData(@"#region ""goo""", true, LanguageNames.CSharp)]
         [InlineData("// File", true, LanguageNames.CSharp)]
         [InlineData("#if DEBUG", false, LanguageNames.VisualBasic)]
+        [InlineData(@"#region ""goo""", true, LanguageNames.VisualBasic)]
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs:868]: woot.  now documents with other pp directives can be reused when changing project configs.  In roslyn this is 4000+ docs.

File: src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs [hunk 2]
```diff
@@ -350,16 +351,20 @@ private DocumentState SetParseOptions(ParseOptions options, bool onlyPreprocesso
...
             if (onlyPreprocessorDirectiveChange &&
                 _treeSource.TryGetValue(out var existingTreeAndVersion))
             {
                 var existingTree = existingTreeAndVersion.Tree;
 
                 SyntaxTree? newTree = null;
 
-                if (existingTree.TryGetRoot(out var existingRoot) && !existingRoot.ContainsDirectives)
+                var syntaxKinds = _languageServices.GetRequiredService<ISyntaxKindsService>();
+                if (existingTree.TryGetRoot(out var existingRoot) && !existingRoot.ContainsDirective(syntaxKinds.IfDirectiveTrivia))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs:367]: utilizes new compiler API we added for this exact purpose.  `.ContainsDirectives` is too broad, firing even for things that do not affect parsing.  `.ContainsDirective(SyntaxKind.IfDirectiveTrivia)` means we only fire on trees hat could genuinely parse differently.

Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs:367]: Is there any broken code scenarios we need to worry about, like there's not an #if but there's an #elif that still results in different trees? I sure don't care about optimizing this case; this is entirely a "I hope we don't crash" question.

**Review by jasonmalinowski — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
> Is there any broken code scenarios we need to worry about, like there's not an #if but there's an #elif that still results in different trees? I sure don't care about optimizing this case; this is entirely a "I hope we don't crash" question.

Tested at compiler layer already.

Note:  `elif` without an `if` produces a `BadDirectiveSyntax`.  That said, even if it didn't, it wouldn't matter as it's fine in error cases for the trees to share that same bad tree.  I'll add tests though to demontrate we don't crash.

**Pull request merged**